src/index.ts

```diff
@@ -1,18 +1,106 @@
 import base64 = require('base-64');
+import queryStringHelpers = require('querystring');
+import urlHelpers = require('url');
+
+const emptyStringToUndefined = (str: string): string | undefined => (str === '' ? undefined : str);
+
+// TS doesn't return the correct type for array and object index signatures. It returns `T` instead
+// of `T | undefined`. These helpers give us the correct type.
+const getIndex = <X>(index: number, xs: X[]): X | undefined => xs[index];
+const getKey = <X>(index: string, xs: { [key: string]: X }): X | undefined => xs[index];
+
+type Maybe<T> = undefined | T;
+
+const isMaybeDefined = <T>(maybeT: Maybe<T>): maybeT is T => maybeT !== undefined;
+
+const mapMaybe = <T, B>(f: (t: T) => B, maybeT: Maybe<T>): Maybe<B> =>
+  isMaybeDefined(maybeT) ? f(maybeT) : maybeT;
+
+const getOrElseMaybe = <T>(f: () => T, maybeT: Maybe<T>): T =>
+  isMaybeDefined(maybeT) ? maybeT : f();
+
+const normalizeMaybe = <T>(nullMaybe: T | null) => (nullMaybe === null ? undefined : nullMaybe);
 
 const TRACKING_PARAM = 'ixid';
 
-const trackingExpression = /(ixid=.*(?=&))|(ixid=.*)/i;
-export const _findTrackingParamsInUrl = (url: string) => {
-  const matches = url.match(trackingExpression);
+const getQueryStringFromParsedUrl = (parsedUrl: urlHelpers.Url): Maybe<string> =>
+  normalizeMaybe(
+    // We cast here to workaround Node typings which incorrectly specify any
+    parsedUrl.query as null | string,
+  );
+type Query = Record<string, string>;
+const parseQueryString = (str: string): Query =>
+  // We cast here to workaround Node typings which incorrectly specify any
+  queryStringHelpers.parse(str) as Query;
+const getTrackingQueryParam = (query: Query) => getKey(TRACKING_PARAM, query);
+
+const omit = <T>(key: string, obj: Record<string, T>): Record<string, T> => {
+  const copy = { ...obj };
+  delete copy[key];
+  return copy;
+};
 
-  if (matches !== null) {
-    return matches[0];
-  } else {
-    return '';
-  }
+const set = <T>(key: string, value: T, obj: Record<string, T>): Record<string, T> => {
+  const copy = { ...obj };
+  copy[key] = value;
+  return copy;
+};
+
+const mapQueryForUrl = (
+  fn: (query: Query) => Query,
+  queryStringifyOptions: queryStringHelpers.StringifyOptions,
+) => (url: string) => {
+  const parsedUrl = urlHelpers.parse(url);
+
+  const maybeQueryString = getQueryStringFromParsedUrl(parsedUrl);
+  const query = getOrElseMaybe(() => ({}), mapMaybe(parseQueryString, maybeQueryString));
+
+  const newQuery: Query = fn(query);
+  const newQueryString = queryStringHelpers.stringify(
+    newQuery,
+    undefined,
+    undefined,
+    queryStringifyOptions,
+  );
+
+  const newParsedUrl = { ...parsedUrl, search: newQueryString };
+  const newUrl = urlHelpers.format(newParsedUrl);
+
+  return newUrl;
+};
+
+const identity = <T>(t: T) => t;
+
+//
+// End generic helpers
```

Comment:
A lot of stuff here that could be imported from external modules, e.g. `identity`, `set`, `omit`, Maybe helpers…

If we do add dependencies, we'll need to make sure they don't have duplication when we later depend on this module in unsplash-web. E.g., I'd like to use funfix's Option, but we'll have to be careful to ensure the versions match. /cc @samijaber

Reply:
Also, if we care about open sourcing this (?), we may want to think twice about adding large dependencies, as it could put users off. Thoughts? @lukechesser

Reply:
@OliverJAsh so yeah that was my thing, was I was trying not to include too many dependencies, even if it means the code being a bit more opaque.

It's not so much because I feel that anyone in the open-source community will want to use this (there may be a few Imgix customers who have our use case), but it's more for our other projects. Some, like Instant, have completely different dependencies, and I'm trying to avoid us forcing them into requiring a specific setup.

That said though, I think there's a balance to be had. As long as the `dist` is consumable and not _too_large, then I think we'll find a good balance.

Reply:
Agreed! We need to strike a balance, to try and avoid re-inventing the wheel with non-battle tested solutions, whilst retaining a small dependency tree. The biggest problem is the lack of a good standard library in JS, which leads to the `set`, `omit`, `getValues` et al helpers you see here. :-(

@samijaber Perhaps we could depend on the individual Lodash modules like `lodash.omit`? As for a Maybe/Option type, funfix's `Option` is 3 KB gzipped, or we can use https://github.com/OliverJAsh/simple-maybe which is basically what I've copied and pasted here. Thoughts?